Auto enable market app on update and improve app update logging

tomneedham opened the pull request:
<!--
Thanks for submitting a change to ownCloud!

This is the bug tracker for the Server component. Find other components at https://github.com/owncloud/core/blob/master/.github/CONTRIBUTING.md#guidelines

For fixing potential security issues please see https://owncloud.org/security/

To make it possible for us to get your change reviewed and merged please fill out below information carefully.

Please note that any kind of change first has to be submitted to the master branch which holds the next major version of ownCloud.

We will carefully discuss if your change can or has to be backported to stable branches.
-->

## Description
 - Enables the market app when upgrading to use it for updating apps and finding apps that are unbundled to the marketplace now but are still in use.

## Related Issue
 - Upgrading to 10.0.1 from 9.x

## Motivation and Context
Otherwise we get this: 
![screen shot 2017-05-18 at 10 28 04](https://cloud.githubusercontent.com/assets/660805/26199908/b748ea1e-3bc3-11e7-97d3-8b8ac420f2cc.png)
which is actually caused by the market app not being enabled


## How Has This Been Tested?
 - tests with 9.1.4 updating to 10.0.x with user_ldap installed and enabled

## Screenshots (if appropriate):
current output with additional messages:
![screen shot 2017-05-18 at 12 09 33](https://cloud.githubusercontent.com/assets/660805/26199956/fc115a00-3bc3-11e7-88c4-c78550d0e82f.png)


## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

# TODO
 - [ ] testing & thinking

tomneedham committed:
Auto enable market app on update and improve app update logging

tomneedham committed:
Display information on why apps fails to update from the market

PVince81 commented on lib/private/Repair/Apps.php:
```diff
@@ -86,45 +90,64 @@ private function isCoreUpdate() {
-		if ($isMarketEnabled && $isCoreUpdate) {
-			$this->loadApp('market');
-			try {
+		if($isCoreUpdate) {
```
also need to check if we're updating from OC < 10.0. We don't want to re-re-re-re-re-enable the app for every minor 10.0 series update if an admin decided to disable it (ex: offline instance)

PVince81 replied:
in this case you still need `$isMarketEnabled` for the OC >= 10.0 case

PVince81 commented:
Not directly related to this PR but let's fix this as well: now when applying this the update moves forward and the market app tries to download the app using OC core's HTTP Client, which itself calls the certificate manager which uses a View which seems to try to use the oc_accounts table:
```
Repair warning: An exception occurred while executing 'SELECT * FROM `oc_accounts` WHERE `lower_user_id` = :dcValue1' with params ["files_external"]:
```

will try and post the full stack once I have it

PVince81 commented:
so far I'm here:
```
0  OC\Security\CertificateManager->createCertificateBundle() /srv/www/htdocs/owncloud-enterprise/lib/private/Security/CertificateManager.php:107
1  OC\Security\CertificateManager->getAbsoluteBundlePath() /srv/www/htdocs/owncloud-enterprise/lib/private/Security/CertificateManager.php:207
2  OC\Http\Client\Client->setDefaultOptions() /srv/www/htdocs/owncloud-enterprise/lib/private/Http/Client/Client.php:73
3  OC\Http\Client\Client->get() /srv/www/htdocs/owncloud-enterprise/lib/private/Http/Client/Client.php:135
4  OCA\Market\MarketService->httpGet() /srv/www/htdocs/owncloud-enterprise/apps/market/lib/MarketService.php:378
5  OCA\Market\MarketService->queryData() /srv/www/htdocs/owncloud-enterprise/apps/market/lib/MarketService.php:411
6  OCA\Market\MarketService->getApps() /srv/www/htdocs/owncloud-enterprise/apps/market/lib/MarketService.php:356
7  OCA\Market\MarketService->getAppInfo() /srv/www/htdocs/owncloud-enterprise/apps/market/lib/MarketService.php:190
8  OCA\Market\MarketService->downloadPackage() /srv/www/htdocs/owncloud-enterprise/apps/market/lib/MarketService.php:117
9  OCA\Market\MarketService->installApp() /srv/www/htdocs/owncloud-enterprise/apps/market/lib/MarketService.php:87
10 OCA\Market\Listener->reinstallAppStoreApp() /srv/www/htdocs/owncloud-enterprise/apps/market/lib/Listener.php:45
11 OCA\Market\Application->OCA\Market\{closure}() /srv/www/htdocs/owncloud-enterprise/apps/market/lib/Application.php:54
12 call_user_func:{/srv/www/htdocs/owncloud-enterprise/lib/composer/symfony/event-dispatcher/EventDispatcher.php:174}() /srv/www/htdocs/owncloud-enterprise/lib/composer/symfony/event-dispatcher/EventDispatcher.php:174
13 Symfony\Component\EventDispatcher\EventDispatcher->doDispatch() /srv/www/htdocs/owncloud-enterprise/lib/composer/symfony/event-dispatcher/EventDispatcher.php:174
14 Symfony\Component\EventDispatcher\EventDispatcher->dispatch() /srv/www/htdocs/owncloud-enterprise/lib/composer/symfony/event-dispatcher/EventDispatcher.php:43
15 OC\Repair\Apps->getAppsFromMarket() /srv/www/htdocs/owncloud-enterprise/lib/private/Repair/Apps.php:180
16 OC\Repair\Apps->run() /srv/www/htdocs/owncloud-enterprise/lib/private/Repair/Apps.php:125
17 OC\Repair->run() /srv/www/htdocs/owncloud-enterprise/lib/private/Repair.php:88
18 OC\Updater->doUpgrade() /srv/www/htdocs/owncloud-enterprise/lib/private/Updater.php:212
19 OC\Updater->upgrade() /srv/www/htdocs/owncloud-enterprise/lib/private/Updater.php:108
20 OC\Core\Command\Upgrade->execute() /srv/www/htdocs/owncloud-enterprise/core/Command/Upgrade.php:249
21 Symfony\Component\Console\Command\Command->run() /srv/www/htdocs/owncloud-enterprise/lib/composer/symfony/console/Command/Command.php:262
22 Symfony\Component\Console\Application->doRunCommand() /srv/www/htdocs/owncloud-enterprise/lib/composer/symfony/console/Application.php:826
23 Symfony\Component\Console\Application->doRun() /srv/www/htdocs/owncloud-enterprise/lib/composer/symfony/console/Application.php:189
24 Symfony\Component\Console\Application->run() /srv/www/htdocs/owncloud-enterprise/lib/composer/symfony/console/Application.php:120
25 OC\Console\Application->run() /srv/www/htdocs/owncloud-enterprise/lib/private/Console/Application.php:160
26 require_once()  /srv/www/htdocs/owncloud-enterprise/console.php:99
27 {main}          /srv/www/htdocs/owncloud-enterprise/occ:11
```

after that it goes to `View::fopen()` on the folder "data/files_external" which would usually contain certificates for external storage connections. And inside that it fails.
It's probably not useful to go deeper.

We need to add a check "needs upgrade" or something in the cert manager and prevent this logic to kick in.

PVince81 commented:
ok, got it... encryption manager trying to do stuff while accessing this folder, finding out whether the folder must be excluded:
```
0  OC\User\Manager->get() /srv/www/htdocs/owncloud-enterprise/lib/private/User/Manager.php:148
1  OC\User\Manager->userExists() /srv/www/htdocs/owncloud-enterprise/lib/private/User/Manager.php:188
2  OC\Encryption\Util->isExcluded() /srv/www/htdocs/owncloud-enterprise/lib/private/Encryption/Util.php:365
3  OC\Files\Storage\Wrapper\Encryption->fopen() /srv/www/htdocs/owncloud-enterprise/lib/private/Files/Storage/Wrapper/Encryption.php:383
4  OC\Files\Storage\Wrapper\Checksum->fopen() /srv/www/htdocs/owncloud-enterprise/lib/private/Files/Storage/Wrapper/Checksum.php:59
5  OC\Files\View->basicOperation() /srv/www/htdocs/owncloud-enterprise/lib/private/Files/View.php:1121
6  OC\Files\View->fopen() /srv/www/htdocs/owncloud-enterprise/lib/private/Files/View.php:966
7  OC\Security\CertificateManager->createCertificateBundle() /srv/www/htdocs/owncloud-enterprise/lib/private/Security/CertificateManager.php:107
8  OC\Security\CertificateManager->getAbsoluteBundlePath() /srv/www/htdocs/owncloud-enterprise/lib/private/Security/CertificateManager.php:207
9  OC\Http\Client\Client->setDefaultOptions() /srv/www/htdocs/owncloud-enterprise/lib/private/Http/Client/Client.php:73
10 OC\Http\Client\Client->get() /srv/www/htdocs/owncloud-enterprise/lib/private/Http/Client/Client.php:135
11 OCA\Market\MarketService->httpGet() /srv/www/htdocs/owncloud-enterprise/apps/market/lib/MarketService.php:378
12 OCA\Market\MarketService->queryData() /srv/www/htdocs/owncloud-enterprise/apps/market/lib/MarketService.php:411
13 OCA\Market\MarketService->getApps() /srv/www/htdocs/owncloud-enterprise/apps/market/lib/MarketService.php:356
14 OCA\Market\MarketService->getAppInfo() /srv/www/htdocs/owncloud-enterprise/apps/market/lib/MarketService.php:190
15 OCA\Market\MarketService->downloadPackage() /srv/www/htdocs/owncloud-enterprise/apps/market/lib/MarketService.php:117
16 OCA\Market\MarketService->installApp() /srv/www/htdocs/owncloud-enterprise/apps/market/lib/MarketService.php:87
17 OCA\Market\Listener->reinstallAppStoreApp() /srv/www/htdocs/owncloud-enterprise/apps/market/lib/Listener.php:45
18 OCA\Market\Application->OCA\Market\{closure}() /srv/www/htdocs/owncloud-enterprise/apps/market/lib/Application.php:54
19 call_user_func:{/srv/www/htdocs/owncloud-enterprise/lib/composer/symfony/event-dispatcher/EventDispatcher.php:174}() /srv/www/htdocs/owncloud-enterprise/lib/composer/symfony/event-dispatcher/EventDispatcher.php:174
20 Symfony\Component\EventDispatcher\EventDispatcher->doDispatch() /srv/www/htdocs/owncloud-enterprise/lib/composer/symfony/event-dispatcher/EventDispatcher.php:174
21 Symfony\Component\EventDispatcher\EventDispatcher->dispatch() /srv/www/htdocs/owncloud-enterprise/lib/composer/symfony/event-dispatcher/EventDispatcher.php:43
22 OC\Repair\Apps->getAppsFromMarket() /srv/www/htdocs/owncloud-enterprise/lib/private/Repair/Apps.php:180
23 OC\Repair\Apps->run() /srv/www/htdocs/owncloud-enterprise/lib/private/Repair/Apps.php:125
24 OC\Repair->run() /srv/www/htdocs/owncloud-enterprise/lib/private/Repair.php:88
25 OC\Updater->doUpgrade() /srv/www/htdocs/owncloud-enterprise/lib/private/Updater.php:212
26 OC\Updater->upgrade() /srv/www/htdocs/owncloud-enterprise/lib/private/Updater.php:108
27 OC\Core\Command\Upgrade->execute() /srv/www/htdocs/owncloud-enterprise/core/Command/Upgrade.php:249
28 Symfony\Component\Console\Command\Command->run() /srv/www/htdocs/owncloud-enterprise/lib/composer/symfony/console/Command/Command.php:262
29 Symfony\Component\Console\Application->doRunCommand() /srv/www/htdocs/owncloud-enterprise/lib/composer/symfony/console/Application.php:826
30 Symfony\Component\Console\Application->doRun() /srv/www/htdocs/owncloud-enterprise/lib/composer/symfony/console/Application.php:189
31 Symfony\Component\Console\Application->run() /srv/www/htdocs/owncloud-enterprise/lib/composer/symfony/console/Application.php:120
32 OC\Console\Application->run() /srv/www/htdocs/owncloud-enterprise/lib/private/Console/Application.php:160
33 require_once()  /srv/www/htdocs/owncloud-enterprise/console.php:99
34 {main}          /srv/www/htdocs/owncloud-enterprise/occ:11
```

let's see if we can mitigate this...

Commit:
Exclude files_external from encryption wrapper

Prevents calling user manager when opening certifications from
certificate manager when using the core http client when attempting to
download a market app during an upgrade from 9.0 where oc_accounts is
not available yet.

PVince81 commented:
Fixed here, excluding "files_external": https://github.com/owncloud/core/pull/27930/commits/9266fef392e62019b8f3c3d895833b63040c7e46

PVince81 commented:
- [ ] TODO: don't auto-enable market if upgrading from OC < 10.0. if market is disabled in this case, keep it disabled and fail only if apps DO require an upgrade through market

tomneedham commented:
- [ ] user_ldap is both compatible & missing and is requested twice from the marketplace

tomneedham committed:
Check if we need to enable market app before force-enabling

PVince81 commented on lib/private/Repair/Apps.php:
```diff
@@ -80,51 +84,90 @@ private function isCoreUpdate() {
-					'upgradeAppStoreApp'
-				);
+				// Try to update incompatible apps
+				if($appsToUpgrade[self::KEY_INCOMPATIBLE]) {
```
should that be `isset()` ? is this test really required ?

PVince81 replied:
`isset()` and `!empty()` ?

PVince81 commented on tests/lib/Repair/AppsTest.php:
```diff
@@ -0,0 +1,68 @@
+ *
+ * @see \OC\Repair\AvatarPermissionsTest
+ */
+class AvatarPermissionsTest extends \Test\TestCase {
```
Avatars?

PVince81 commented:
@tomneedham see my comments. Some minor things and avatars

PVince81 commented on tests/lib/Repair/AppsTest.php:
```diff
@@ -0,0 +1,68 @@
+ *
+ * @group DB
+ *
+ * @see \OC\Repair\AvatarPermissionsTest
```
the avatar is still there, and so is the author...

tomneedham replied:
🙈

PVince81 commented on tests/lib/Repair/AppsTest.php:
```diff
@@ -0,0 +1,68 @@
+/**
+ * @author Roeland Jago Douma <rullzer@owncloud.com>
+ *
+ * @copyright Copyright (c) 2016, ownCloud, Inc.
```
what year is this ?

PVince81 commented:
:+1:

tomneedham commented:
Seems something went weird with docker in jenkins: 

```
16:55:49 Waiting for MySQL(utf8mb4) initialisation ...
17:05:50 ................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................
17:05:50 [ERROR] Waited 600 seconds, no response
17:05:50 Kill the docker adab554079d8ef09d2081be106de042dce91b7e6e06734078fa2513676902a47
17:05:50 adab554079d8ef09d2081be106de042dce91b7e6e06734078fa2513676902a47
17:05:50 adab554079d8ef09d2081be106de042dce91b7e6e06734078fa2513676902a47
17:05:50 Makefile:179: recipe for target 'test-php' failed
17:05:50 make: *** [test-php] Error 1
[Pipeline] }
[Pipeline] // timeout
[Pipeline] echo
17:05:50 Test execution failed: hudson.AbortException: script returned exit code 2
[Pipeline] step
17:05:50 Recording test results```

PVince81 commented:
> Seems something went weird with docker in jenkins:

probably only the 34598th time it's happening, just rebase to retry... and bring a back of luck!

tomneedham commented:
rebased

tomneedham commented:
🙏 for jenkins
